src/ucp/tag/eager_snd.c

```diff
@@ -253,10 +253,19 @@ static ucs_status_t ucp_tag_eager_sync_bcopy_multi(uct_pending_req_t *self)
 void
 ucp_tag_eager_sync_zcopy_req_complete(ucp_request_t *req, ucs_status_t status)
 {
+    ucs_assert(req->send.state.uct_comp.count == 0);
+    ucp_request_send_buffer_dereg(req); /* TODO register+lane change */
+    ucp_tag_eager_sync_completion(req, UCP_REQUEST_FLAG_LOCAL_COMPLETED,
+                                  status);
+}
+
+void ucp_tag_eager_sync_zcopy_completion(uct_completion_t *self,
+                                         ucs_status_t status)
+{
+    ucp_request_t *req = ucs_container_of(self, ucp_request_t,
+                                          send.state.uct_comp);
     if (req->send.state.dt.offset == req->send.length) {
-        ucp_request_send_buffer_dereg(req); /* TODO register+lane change */
-        ucp_tag_eager_sync_completion(req, UCP_REQUEST_FLAG_LOCAL_COMPLETED,
-                                      status);
+        ucp_tag_eager_sync_zcopy_req_complete(req, status);
```

Comment:
this fix seems weird. if uct_am_zcopy, returns UCS_OK, ucp_tag_eager_sync_zcopy_completion should never be called

Reply:
yes, right
`ucp_tag_eager_sync_zcopy_req_complete` is called in this case

Reply:
`ucp_tag_eager_sync_zcopy_completion` is called from `ucp_request_t::uct_completion_t::func` callback when Zcopy operation is completed

Reply:
it shouldn't be called if uct_am_zcopy returns OK
maybe it's an issue with TCP implementation for zcopy

Reply:
discussed offline with @yosefe 

by this:

> yes, right
> `ucp_tag_eager_sync_zcopy_req_complete` is called in this case

I meant that this is called from https://github.com/openucx/ucx/blob/1e227c972f1724d02751f65ba6196ea9d29a3349/src/ucp/proto/proto_am.inl#L197

Reply:
yes, the problem is that complete() is called before dt.offset is updated, so it fails to complete UCP request.